### PR TITLE
Mobile product listing tweaks

### LINKED ISSUE #4598: [mobile ux] Add products to the cart redesign
## In a feature branch or master?
Product listing feature branch 

## Description
**- As a:** shopper
**- On page:** the product list in a shop
**- I want to be able to do:** quickly add or remove products in my cart

## Acceptance Criteria & Tests
- [ ] Standard product listing

**HOW IT LOOKS NOW**
Screen Shot 2019-12-18 at 2 30 47 pm
Screen Shot 2019-12-18 at 2 31 06 pm

**HOW IT NEEDS TO LOOK**
Screen Shot 2019-12-18 at 2 31 40 pm
Screen Shot 2019-12-18 at 2 31 46 pm

- [ ] Group buy min/max product listing

**HOW IT LOOKS NOW**
On mobile:
Screen Shot 2020-05-26 at 5 26 52 pm

On desktop:
Screen Shot 2020-05-26 at 5 26 25 pm

**HOW IT NEEDS TO LOOK**
On mobile:
OFN Shop Product List Bulk Order 20200518

OFN Shop Product List Bulk Order Modal 20200518

On desktop:
Desktop OFN Shop Product List Bulk Order 20200518

OFN Desktop Shop Product List Bulk Order Modal 20200518

------
### Testing notes
**Test single products**

ADD A PRODUCT TO CART
1. View product list on a shop
2. Click the add button on a product
3. That product should be added to the cart with a quantity of 1; the cart icon at top of page should show a '1'; the 'add' button should change to be a plus/minus buttons; '1 in cart' copy should display under buttons.

INCREASE QUANTITY IN CART FOR A PRODUCT
1. Click the '+' button on a product
2. The cart should now have 2 of that product; the cart icon at the top of the page should show a '2'; '2 in cart' copy should display under buttons.

That product should be added to the cart with a quantity of 1; the cart icon at top of page should show a '1'; the 'add' button should change to be a plus/minus buttons; '1 in cart' copy should display under buttons.

DECREASE QUANTITY IN CART FOR A PRODUCT
1. Click the '-' button on a product
2. The cart should now have 1 of that product; the cart icon at the top of the page should show a '1'; '1 in cart' copy should display under buttons.

**Test bulk products**
ADD A BULK PRODUCT TO CART
1. View product list on a shop
2. Click the add button on a product
3. Bulk product add modal displays
4. Click add 1 for min and 2 for max
5. That product should be added to the cart with a quantity of min 1 and max 2; the cart icon at top of page should show a '1'; the min and max numbers should display (as per screenshot) and 'in cart' copy should display under buttons.

INCREASE QUANTITY IN CART FOR A BULK PRODUCT
1. Click on the min/max numbers
2. Bulk product add modal displays
3. Click add 2 for min and 3 for max
4. That product should have a cart quantity of 2, and the min and max numbers should have changed to be 2/3.

Design details are kept in our Zeplin account: https://app.zeplin.io/project/59532764a6c78ff3c534b372/dashboard

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #4598

Most of the new product listing work has been done already but Yuko found a bunch of shortcomings in the implementation. They are hopefully all addressed in this pull request.

products-mobile-to-desktop

#### What should we test?

See issue #4598.

You will need a developer to adjust the product image sizes on your chosen staging server before you can properly test the new layout.

- Use openfoodfoundation/ofn-install#651 to provision the staging server
- Use openfoodfoundation/ofn-install#651 to deploy this pull request.
- SSH to the server and run:
  ```sh
  cd ~/apps/openfoodnetwork/current/
  bundle exec rake paperclip:refresh:metadata CLASS=Spree::Image
  bundle exec rake paperclip:refresh:thumbnails CLASS=Spree::Image
  ```

Original design:
https://app.zeplin.io/project/59532764a6c78ff3c534b372/screen/5e129ab052a96b1b1cf8916e

Testing comments:
https://docs.google.com/document/d/1WZZSuMCLHiZ_UY_nLJZ8MI-HGadTQmVty7zXeS7H6VE

#### Release notes

The product listing in shop fronts has been re-designed. Images now appear bigger and on mobile devices. Adding to cart now works by clicking or tapping instead of entering a quantity.

Changelog Category: Changed

#### Dependencies

After deployment, we need to resize all images for the product listing as described in the testing section. We should solve #5911 at the same time.

#### Documentation updates

The user guides needs some updated screenshots.

### CODE REVIEW

**Comment by luisramos0** [comment]
re #, did you see issue 2 here Maikel:
https://github.com/openfoodfoundation/openfoodnetwork/issues/5911

File: app/assets/stylesheets/darkswarm/_shop-product-thumb.scss [hunk 1]
```diff
@@ -2,9 +2,16 @@
   products {
     product {
       .product-thumb {
-        width: 7rem;
-        height: 7rem;
+        // Desktop: the product summary is nine columns wide. Use two
+        // for the image. In our design, that's a max of 192px.
+        width: calc(100% / 9 * 2);
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/_shop-product-thumb.scss:7]: I'd move the calculation to the comment and just have width: 22.2%

File: app/assets/stylesheets/darkswarm/_shop-product-thumb.scss [hunk 2]
```diff
@@ -2,9 +2,16 @@
...
         float: left;
+
+        // Mobile: the summary has full twelve columns and the image
+        // should take four of them. Maximum is 227px.
+        @include breakpoint(phablet) {
+          width: calc(100% / 12 * 4);
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/_shop-product-thumb.scss:13]: I'd move the calculation to the comment and just have width: 33.3%

File: app/assets/stylesheets/darkswarm/_shop-product-thumb.scss (hunk 1 above)
Comment (luisramos0) [reply app/assets/stylesheets/darkswarm/_shop-product-thumb.scss:7]: I think this part is not relevant and could be removed:
"In our design, that's a max of 192px."

File: app/assets/stylesheets/darkswarm/_shop-product-thumb.scss (hunk 2 above)
Comment (luisramos0) [reply app/assets/stylesheets/darkswarm/_shop-product-thumb.scss:13]: I think this is not relevant and could be removed?
"Maximum is 227px."

File: app/assets/stylesheets/darkswarm/_shop-product-rows.scss [hunk 3]
```diff
@@ -97,9 +94,14 @@
         }
 
         .summary-header {
-          padding-left: 7.9375rem;
+          // product-thumb width + 1rem
+          padding-left: calc(100% / 9 * 2 + 1rem);
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/_shop-product-rows.scss:98]: I dont see the need to complicate here.
I'd make it
calc(22.2% + 1rem);

File: app/assets/stylesheets/darkswarm/_shop-product-rows.scss [hunk 4]
```diff
@@ -97,9 +94,14 @@
...
           padding-right: 1rem;
 
+          @include breakpoint(phablet) {
+            padding-left: calc(100% / 12 * 4 + 1rem);
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/_shop-product-rows.scss:102]: same here, I'd make it
calc(33.3% + 1rem);

File: app/assets/stylesheets/darkswarm/_shop-product-rows.scss [hunk 5]
```diff
@@ -15,7 +15,8 @@
       }
 
       .shop-variants {
-        padding-left: 7.9375rem;
+        // product-thumb width + 1rem
+        padding-left: calc(100% / 9 * 2 + 1rem);
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/_shop-product-rows.scss:19]: same here
I'd make it
calc(22.2% + 1rem);

File: app/assets/stylesheets/darkswarm/shop.scss [hunk 6]
```diff
@@ -67,7 +71,13 @@
     product {
       @include csstrans;
 
-      border-top: 1px solid #e5e5e5;
+      // Avoid margin collapsing which breaks the flush alignment of the first
+      // image. https://stackoverflow.com/a/19719427/3377535
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/shop.scss:75]: I dont think the SO article is specific enough to deserve to be here.
I'd keep the comment only, any dev can google "avoid margin collapse" and find that article or similar.

File: app/assets/stylesheets/darkswarm/shop.scss [hunk 7]
```diff
@@ -67,7 +71,13 @@
...
+      overflow: auto;
+
+      &:nth-child(n + 2) {
+        border-top: 1px solid #e5e5e5;
```
Comment (luisramos0) [inline app/assets/stylesheets/darkswarm/shop.scss:79]: this color is $disabled-light

**Review by luisramos0 — APPROVED** [review]
Awesome :-)

See comments below.

**Comment by Erioldoesdesign** [comment]
Look at those scaling images :D amazing

**Comment by luisramos0** [comment]
@mkllnk can you add details about what exactly a dev needs to run in staging and if there's something we need to do in prod upon deployment. Thanks!

File: app/assets/stylesheets/darkswarm/shop.scss (hunk 7 above)
Comment (mkllnk) [reply app/assets/stylesheets/darkswarm/shop.scss:79]: Well spotted. That was actually the wrong colour. I adjusted it.

**Review by luisramos0 — APPROVED** [review]
:ok_hand:

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by Matt-Yorkley** [comment]
> bundle exec rake paperclip:refresh:thumbnails CLASS=Spree::Image

We need to think about how to handle this when deploying it to production. Also needs careful testing in staging; with servers that use S3 and servers that don't use S3.

**Comment by luisramos0** [comment]
/admin/image_settings/edit.
Under Styles Paperclip set the Small style to 227x227#.


oh, I didnt realize this was changing with this PR :warning:
Image Settings is about to be deleted here: https://github.com/openfoodfoundation/openfoodnetwork/pull/5729
And ofn-install is already overritting these values on the database: https://github.com/openfoodfoundation/ofn-install/pull/645/files

Because 645 is already merged, I think your instructions should not be followed and we should instead have a PR in ofn-install to change the values defined in PR 645. After that, we still need to run the refresh thumbnails command.
If we dont do this, the changes done in image settings will be overwritten in the following deploy with the values defined in 645.

I am moving this to In Dev. Please let me know if this makes sense to you.

**Comment by luisramos0** [comment]
If we want to go with # instead of > for mini and small then we can make all instances use the same sizes, that means this would be the PR:
https://github.com/openfoodfoundation/ofn-install/pull/651/files

**Comment by Matt-Yorkley** [comment]
We'll want to do something like:

- Update image settings
- Deploy the release
- Refresh all thumnails

...for each instance, as we're deploying it. Right?

**Comment by mkllnk** [comment]
@daniellemoorhead raised that ideally we would resize the images before deploying the new code. So when the new design goes online, it looks perfect straight away. We could deploy the new image formats with https://github.com/openfoodfoundation/ofn-install/pull/651 next Tuesday and the new design could go online the week after (if testing doesn't reveal anything big).

**Comment by mkllnk** [comment]
@luisramos0 I changed the testing instructions to use your PR for the image settings. With that, I moved this to test-ready and added your PR as dependency (blocked by). This shouldn't get merged before your PR but it would be great to get the testing feedback to know that the image size is right. I'm also expecting more feedback about little optimisations and that could delay the mobile work again.

**Comment by luisramos0** [comment]
ok Maikel :+1: 
with that deployment strategy we need to test the current design with 651: is the current product list not totally broken with 227px images? According to your plan this would be live for at least one week.

**Comment by mkllnk** [comment]
> is the current product list not totally broken with 227px images?

No, I tested that. The current design displays the images as 112x112 even though the file is bigger.

**Comment by mkllnk** [comment]
I tried to follow the staging instructions and failed miserably. Over time, people uploaded invalid images or some images are lost. And the resize task just crashes when encountering an error like that. I tried to fix it but then just encountered other errors. I think we need to write our own task that handles errors better and maybe removes broken images.

I remember that I had to delete some broken images when running this task on au production a while ago. But staging is much worse, there are many more broken images.

**Comment by daniellemoorhead** [comment]
@mkllnk are you able to get this back up on staging for @JenRSmith to complete testing?

**Comment by RachL** [comment]
@daniellemoorhead @filipefurtad0 I don't know which one of you will test this, but it could be interesting to include a look at the bulk buy option and check if we correctly fix this issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/5854 and otherwise reopen the issue. Thanks :)

**Comment by daniellemoorhead** [comment]
Jen is testing this at the moment.

**Comment by mkllnk** [comment]
> are you able to get this back up on staging for @JenRSmith to complete testing?

Yes, I'm trying to. Lots of boulders in the way but it may actually work today.

**Comment by mkllnk** [comment]
Okay, the pull request is still deployed and I fixed the image paths. Testing can continue...

I'll ping @JenRSmith on Slack.

**Comment by JenRSmith** [comment]
@RachL I tested the bulk buy option (iPhone/Safari & Windows/Chrome) & it has been fixed. It now looks like this - 
bulk_buy_option

**Comment by RachL** [comment]
@JenRSmith thank you so much! this looks awesome 😎

**Comment by JenRSmith** [comment]
As per my discussion with @mkllnk, there is a problem with the price breakdown icon. When there are multiple, the subsequent icons (2nd, 3rd etc.) are no longer clickable once one of the icons has been clicked. Any can be clicked initially, but only the first icon can be clicked beyond the initial click. I found this when testing on  iPhone 7/Safari & Windows 10/Chrome.

https://drive.google.com/file/d/10lJp6wS-4_9SRUFRucjwzXFmKtRWEFmt/view?usp=sharing

**Comment by mkllnk** [comment]
Thank you for all the testing! There's still quite a bit to do. So I will merge this into the feature branch (successfully reviewed) and start a new pull request with the outstanding issues. That will make the code review easier.

@daniellemoorhead The testing doc contains many rows which have an empty "Fixed?" column. What does that mean? Are those things not tested yet?

### Maybe untested issues:

#### Desktop

- [ ] Product name h3 font size is too large (especially on mobile).
- [ ] Product name h3 margins need a little tweaking to improve the overall spacing.
- [ ] Padding between Product description and Producer name/link is too large.
- [ ] Padding below weight, price breakdown, cost, and add button row is too large.
- [ ] When you add a product to your cart, the padding issue above decreases… Button actions should not affect padding/margins.
- [ ] Product thumbnail images are too small.

#### Tablet

- [ ] Unit cost and Total cost need to be moved to the right on tablet.


### Outstanding issues

#### Desktop

- [x] Pie button for price breakdown not always clickable.
- [x] Pie button not aligned with product name.
- [ ] First product listing should be flush with the search field row (currently there is a gap).
- [ ] Not sure what is the condition that is causing the ‘1 30cm pie’ text in grey to appear here below the first variant.
- [ ] Remove hover state from product tags.

#### Tablet

- [ ] Not sure why this happens but in mobile & tablet the product descriptions here have white spaces in front of them.

#### Mobile

- [ ] Not sure why this happens but in mobile & tablet the product descriptions here have white spaces in front of them.
- [ ] Variant names are not displaying if they aren’t different to the product name...however now it looks like numbers are floating in space.

**Comment by daniellemoorhead** [comment]
>@daniellemoorhead The testing doc contains many rows which have an empty "Fixed?" column. What does that mean? Are those things not tested yet?

@mkllnk the testing doc has green rows for those that have been tested and are now passing, and red rows where they either have been tested and still fail, or @yukoosawa is yet to complete her testing on them. I'm not sure that the list you have above is actually correct, or at least is comprehensive given the incomplete testing status.

@yukoosawa is slated to do her testing this weekend. We should have a complete view of what's left at that point in time. Yuko, would suggest that we create a dot point list on here of the outstanding items to be done (the testing doc feels like it's getting confusing, though if you can make it less so be my guest 😄 )

**Comment by JenRSmith** [comment]
@mkllnk have tested those 5 items listed above and all pass.

**Comment by daniellemoorhead** [comment]
@mkllnk all that's left to do is merge this PR to your product branch and then we can close this card and keep #6080 as the only remaining work to be done for the product listing epic.

**Pull request merged**